REGRESSION (iOS 16): WebGL 2 crashing on glReadPixels

kkinnunen-apple opened the pull request:
#### 5192197861cfe329db2f05ec91f887d11f7b3da3
<pre>
REGRESSION (iOS 16): WebGL 2 crashing on glReadPixels
<a href="https://bugs.webkit.org/show_bug.cgi?id=245476">https://bugs.webkit.org/show_bug.cgi?id=245476</a>
rdar://problem/100252324

Reviewed by Alex Christensen.

ReadPixels is currently a method that sends the client
passed data buffer to GPUP, reads the pixels into that
and passes the data back to WP.
Conceptually message of form ReadPixels(uint8_t* data) -&gt; (uint8_t* data)

Before:
IPC encoding of memory arrays for calls like ReadPixels
would work as follows:
 - If the array fits to the stream connection command buffer,
   copy the array there
 - Alternatively downgrade the message to out-of-stream message.
   - Copy the array to the IPC::Encoder message body buffer.
   - Pass the message body as Mach message memory mapping.

IPC decoding of memory arrays for calls like ReadPixels
would work as follows:
 - Allocate temp Vector for the data
 - Copy the data from stream command buffer / Mach message memory
   mapping into the temp buffer
 - Call ReadPixels
 - Copy the temp Vector to IPC::Encoder message body buffer
   similar to the encoding phase, and send it back to WP.

This would have two problems:
 - Extensive copying is slow for very big buffers
 - The temp Vector and Mach message buffers via IPC encoding
   contribute to GPUP footprint.
   - The call IPC encoding buffer is not explicitly attributed to
     the caller. It&apos;s unclear to which process this belongs, likely
     this contributes to GPUP footprint.
   - The temp Vector is not attributed to the caller, contributing
     to GPUP footprint.
   - The reply IPC encoding buffer is not attributed to the caller,
     contributing to the GPUP footprint.
   - The ownership of the reply IPC encoding buffer, when passed as
     Mach message memory, is unclear. Likly this contributes to
     GPUP footprint. (Same as the encoding part).

After:
If the read pixels buffer is large, copy it to a newly allocated
shared memory buffer and use that to read the pixels.
Add a new IPC message variant for passing the shared memory area.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::readnPixels2):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::readnPixels):
(WebKit::RemoteGraphicsContextGLProxy::readnPixelsSharedMemory):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Tools/Scripts/generate-gpup-webgl:

Canonical link: <a href="https://commits.webkit.org/258127@main">https://commits.webkit.org/258127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb927a5859285415bc9b525e763ff239c4976994

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33904 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110161 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170435 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/919 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93269 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108006 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91530 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34886 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90182 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77863 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3698 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24450 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/874 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9827 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43945 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5580 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5494 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
```diff
@@ -53,6 +53,7 @@ messages -> RemoteGraphicsContextGL NotRefCounted Stream {
     void SimulateEventForTesting(WebCore::GraphicsContextGL::SimulatedEventForTesting event)
     void ReadnPixels0(int32_t x, int32_t y, int32_t width, int32_t height, uint32_t format, uint32_t type, IPC::ArrayReference<uint8_t> data) -> (IPC::ArrayReference<uint8_t> data) Synchronous
     void ReadnPixels1(int32_t x, int32_t y, int32_t width, int32_t height, uint32_t format, uint32_t type, uint64_t offset)
+    void ReadnPixels2(int32_t x, int32_t y, int32_t width, int32_t height, uint32_t format, uint32_t type, WebKit::SharedMemory::Handle handle) -> (bool hadError) Synchronous NotStreamEncodable
```
These could use a better name.

kkinnunen-apple committed:
REGRESSION (iOS 16): WebGL 2 crashing on glReadPixels
https://bugs.webkit.org/show_bug.cgi?id=245476
rdar://problem/100252324

Reviewed by Alex Christensen.

ReadPixels is currently a method that sends the client
passed data buffer to GPUP, reads the pixels into that
and passes the data back to WP.
Conceptually message of form ReadPixels(uint8_t* data) -> (uint8_t* data)

Before:
IPC encoding of memory arrays for calls like ReadPixels
would work as follows:
 - If the array fits to the stream connection command buffer,
   copy the array there
 - Alternatively downgrade the message to out-of-stream message.
   - Copy the array to the IPC::Encoder message body buffer.
   - Pass the message body as Mach message memory mapping.

IPC decoding of memory arrays for calls like ReadPixels
would work as follows:
 - Allocate temp Vector for the data
 - Copy the data from stream command buffer / Mach message memory
   mapping into the temp buffer
 - Call ReadPixels
 - Copy the temp Vector to IPC::Encoder message body buffer
   similar to the encoding phase, and send it back to WP.

This would have two problems:
 - Extensive copying is slow for very big buffers
 - The temp Vector and Mach message buffers via IPC encoding
   contribute to GPUP footprint.
   - The call IPC encoding buffer is not explicitly attributed to
     the caller. It's unclear to which process this belongs, likely
     this contributes to GPUP footprint.
   - The temp Vector is not attributed to the caller, contributing
     to GPUP footprint.
   - The reply IPC encoding buffer is not attributed to the caller,
     contributing to the GPUP footprint.
   - The ownership of the reply IPC encoding buffer, when passed as
     Mach message memory, is unclear. Likly this contributes to
     GPUP footprint. (Same as the encoding part).

After:
If the read pixels buffer is large, copy it to a newly allocated
shared memory buffer and use that to read the pixels.
Add a new IPC message variant for passing the shared memory area.

* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::readnPixels2):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::readnPixels):
(WebKit::RemoteGraphicsContextGLProxy::readnPixelsSharedMemory):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Tools/Scripts/generate-gpup-webgl:

Canonical link: https://commits.webkit.org/258127@main
